[buteo-sync-plugins-social] Implement rudimentary connection-loss handling. Contributes to MER#885

chriadam opened the pull request:
This commit adds basic handling for connection loss during sync.
In most cases, it detects the connection loss and then skips any
database commit during finalisation.
In some cases where this is not possible, it merely aborts proceeding
operations and marks the account for clean sync on next cycle.

chriadam commented:
@antseppa @pvuorela

chriadam commented:
Also https://github.com/nemomobile/buteo-sync-plugin-caldav/pull/51 and https://github.com/nemomobile/buteo-sync-plugin-carddav/pull/14

chriadam commented:
Hrm, actually, this only solves half the problem.  Unfortunately, the Google Calendar sync adapter, in "clean sync" codepath, currently commits the removal of all data to the database, prior to beginning the next one....

chriadam commented:
Updated so that the clean sync "remove everything" code doesn't call m_storage->save() by changing the CleanUpPurge to SyncPurge (which ensures that it doesn't commit+close the database).

However, in my testing, it still didn't resolve the issue (ie, it seemed that despite the fact that the code flow works correctly, triggering a clean sync and then disconnecting wifi after the purge finishes but before the sync completes, causes all calendar data for the account to be removed from the calendar)...  I don't understand why.

chriadam commented:
I believe the reason is that mkcal's extendedstorage.cpp deleteNotebook() code by default removes the notebook from storage as well as memory.  It does have a parameter which may be passed to avoid committing to storage, but I have yet to test the results of specifying that - it may have unintended consequences.

chriadam commented:
Specifying the "OnlyMemory" flag is a bad idea - the deletion of the notebook is still committed to storage even when you specify that flag, but the events which were contained in that notebook are NOT deleted from storage.  This results in a large number of "ghost events" being left on the device.

chriadam commented:
Ok... this one is now ready for review and testing.

In the google calendar sync adapter I now try to do all local-database modifications within the scope of a single function (finishedRequestingRemoteEvents()).  If the sync is aborted prior to that, no local database modifications will be performed.  If the sync is aborted after that (eg, during the upsync of the local changes to the remote server) then the local-database-changes will still be committed to the local database, but the sync will be marked as "failed" (so the next sync will be a "clean sync" and just retrieve whatever is on the server.

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -727,6 +764,7 @@ void GoogleCalendarSyncAdaptor::calendarsFinishedHandler()
                     currCalendarInfo.color = currCalendar.value(QStringLiteral("backgroundColor")).toString();
                     currCalendarInfo.summary = currCalendar.value(QStringLiteral("summary")).toString();
                     currCalendarInfo.description = currCalendar.value(QStringLiteral("description")).toString();
+                    currCalendarInfo.change = NoChange; // we detect the appropriate change type (if required) later.
```
This would be already set in ctor?

chriadam replied:
Yes, but I wanted to set it explicitly here just as documentation, basically.

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -592,6 +598,40 @@ void GoogleCalendarSyncAdaptor::finalCleanup()
+
+    // Delete any events which are not associated with a notebook.
+    // These events are ghost events, caused by a bug which previously
+    // existed in the sync adapter code due to mkcal deleteNotebook semantics.
```
Has the bug been in some released version? I understood some ghost events were resulting from earlier version of this PR.

chriadam replied:
See JB#22313 and JB#26514 and JB#27888 - sadly it seems that this is a case which can occur somehow in released software (I still don't understand precisely how).

pvuorela replied:
In that case, I suppose the reference to deleteNotebook is slightly misleading. Wonder if earlier missing recurrence id support could be related to these ghost events. I recall events with recurrence id but missing main event are currently undeletable from calendar.

I guess this is okayish db clean up, but doing it on every sync seems a bit much. Once per device could be enough already. Not fully objecting this version, though.

chriadam replied:
Added code to ensure we only do the orphan cleanup check once (uses gcal.ini key to determine whether to perform the check or not, so can be triggered by removing that file (or setting the key there to false).

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -955,34 +998,163 @@ void GoogleCalendarSyncAdaptor::eventsFinishedHandler()
+    return mKCal::Notebook::Ptr();
+}
+
+void GoogleCalendarSyncAdaptor::finishedRequestingRemoteEvents(int accountId, const QString &accessToken, const QString &calendarId, const QDateTime &since, const QString &updateTimestampStr)
```
For readability would suggest to chop this long lines.

chriadam replied:
ok, will do

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -955,34 +998,163 @@ void GoogleCalendarSyncAdaptor::eventsFinishedHandler()
+                    m_storage->allIncidences(&incidenceList, notebook->uid());
+                    foreach (KCalCore::Incidence::Ptr incidence, incidenceList) {
+                        m_calendar->deleteIncidence(m_calendar->incidence(incidence->uid(), incidence->recurrenceId()));
+                    }
```
Btw, looks to me that deleting and loading the incidences is unnecessary. allIncidences() appears to load, and deleteNotebook() appears to call that plus delete all incidences.

chriadam replied:
Does the deleteNotebook code delete things from calendar?  I had found that calling calendar->deleteIncidence(someIncidenceLoadedFromStorage); // fails

You have to explicitly load an incidence into the calendar in order to delete it properly.  If you don't, you just get a ghost event appearing in the database (ie, notebook is deleted, and incidence has the "notebookUid" field cleared, but the event still exists in the database).

pvuorela replied:
There's the code:
https://github.com/mer-packages/mkcal/blob/master/src/extendedstorage.cpp#L417
https://github.com/mer-packages/mkcal/blob/master/src/sqlitestorage.cpp#L2521

At least looks like doing much of the same things, including calendar()->deleteIncidence( toDelete ), and that it tries to delete even the unloaded incidences. But haven't tested myself, just a remark.

chriadam replied:
Let me test again, I may have been doing something strange when I saw the ghost events appear, or misattributed the cause.

chriadam replied:
I tested this again tonight.  The incidences are definitely not removed from the database unless I delete them manually with the loop above.  I don't know why not, but they certainly don't get purged from the database.

chriadam replied:
Agh, wait, of course they don't get purged... they get marked as deleted instead.  I will re-test to check whether that occurs properly.

chriadam replied:
Ok, you were correct - it does delete them.  However, it turns out that if a notebook is marked readonly, then the deletion of the events fails, leaving ghost events.

I've now updated by code to do two things:
1) remove the "load all incidences" thing, in favour of just deleteNotebook
2) ensure I manually setIsReadOnly(false) prior to the deleteNotebook call.

pvuorela commented on src/google/google-calendars/googlecalendarsyncadaptor.cpp:
```diff
@@ -763,32 +807,32 @@ void GoogleCalendarSyncAdaptor::updateLocalCalendarNotebooks(int accountId, cons
+                    // we are performing a clean sync cycle.
+                    // we will eventually delete and then insert this notebook.
+                    SOCIALD_LOG_DEBUG("queuing clean sync of local calendar" << notebook->name() << currDeviceCalendarId << "for Google account:" << accountId);
+                    m_serverCalendarIdToCalendarInfo[accountId][currDeviceCalendarId].change = GoogleCalendarSyncAdaptor::CleanSync;
```
needCleanSync == !wasLastSyncSuccesfull(). So here local modifications will be gone after two network errors?

As another detail, I'm not sure I fully follow the need to postpone notebook modifications. Here full server side notebook list should be known, so are there bad side-effects if notebook changes are done even if processing events is later aborted?

chriadam replied:
Yes, local modifications will be dropped due to the clean-sync semantics, currently.  Is this a bad idea?  I don't really know how to recover properly from error state after database writes have started, perhaps I'm overlooking something.

For the second thing: the problem is that notebook modifications are immediately flushed to storage (and if we pass the MemoryOnly bool flag, then bad things happen - ghost events everywhere.  I haven't investigated why, I just assume that it's something mkcal specific).  So, if we apply the notebook modifications immediately now, but then the very next network request fails due to connection error, then it's possible that the notebooks will all have been deleted and re-created (due to clean-sync semantics), and have no content / events in them.  Ie, JB#26988

pvuorela replied:
Well, the clean sync at least sounds relatively easy to get. I'm not still familiar with many aspects of sync stack, so can be myself overlooking, but in case sync is aborted because of network error (or similar) wouldn't it be possible to just avoid updating last sync time stamp, and avoiding marking next sync to be a clean one? Maybe even postponing the time stamp update until all changes are pushed too. Next sync would then proceed with time stamp of last fully successful one. If there were partial updates in between, these would just update to same data, i.e. make no changes(?)

On second, maybe I now understood: so if there's a case where clean sync is needed, it's better to be done only if full repopulation of the recreated notebook is possible. Makes sense. Though if clean sync could be avoided more, not so crucial anymore, I guess.

chriadam replied:
Right, there are definitely things we can do to reduce the number of clean-syncs which occur as opposed to delta syncs (especially since there are less cases which can cause database corruption now).

But, IMO, unless it proves to be a common occurrence in practice, I think it's a lower priority issue.  The way we currently store sync timestamps (in an out-of-band database) is terrible but it's that way for historical reasons.  At some stage we need to refactor the plugins to get rid fo the remaining "multiple account" isms, and get rid of the false separation between SocialNetworkSyncAdaptor and Buteo::ClientPlugin, so that we can make "proper" use of the Buteo API within our sync adaptors (ie: return the correct error codes, instead of just InternalError; make use of the Buteo sync-anchors timestamp database; make proper use of the Buteo SyncResults struct to inform whether we need to do cleansync or not; etc).

But, not today :-)

pvuorela replied:
Fair enough.

pvuorela commented:
Suppose the commit message should refer to MER#885?

chriadam commented:
Yes, fixed to refer to MER#885 instead of nonexistent JB#885 - nice catch, thanks!

pvuorela commented:
Apart some small details which are fine this way too, google calendar side looks good to me.

chriadam commented:
Updated - now only does orphan cleanup once, also optimises the deleteNotebook code.  I also fixed a bug in the clean-sync codepath which caused a notebook to be created twice when clean-sync is triggered (can be tested by rm'ing the gcal.ini file in between syncs).

pvuorela commented:
Calendar side still looks good to me. Others too on the surface, but I'm not very familiar with those.
